### PR TITLE
feat(datatable): allow row drag and drop

### PR DESCRIPTION
**Pull request by chintankavathia**
**What kind of change does this PR introduce?** (check one with "x")

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

**What is the current behavior?** (You can also link to an open issue here)

**What is the new behavior?**

**Does this PR introduce a breaking change?** (check one with "x")

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

**Comment by chintankavathia** [comment]
@kfenner This is now ready for review. Please have a look.

File: projects/ngx-datatable/src/lib/types/drag-events.type.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+export type DragEventType = 'DRAG' | 'DRAG_END' | 'DRAG_ENTER' | 'DRAG_LEAVE' |
+'DRAG_OVER' | 'DRAG_START' | 'DROP';
```
Comment (kfenner) [inline projects/ngx-datatable/src/lib/types/drag-events.type.ts:2]: @chintankavathia Could you get these from an existing typescript type or even from the native one? Otherwise, at least make them lower case like this as I don't see the UPPER_CASE being used in ngx-translate.
```suggestion
export type DragEventType = 'drag' | 'drag-end' | 'drag-enter' | 'drag-leave' |
'drag-over' | 'drag-start' | 'drop';
```

Comment (chintankavathia) [reply projects/ngx-datatable/src/lib/types/drag-events.type.ts]: there is no typescript type and native one is used as string.

Comment (kfenner) [reply projects/ngx-datatable/src/lib/types/drag-events.type.ts]: @chintankavathia Thanks, I like that you aligned it with the naming of the native events.

File: projects/ngx-datatable/src/lib/components/body/body.component.ts [hunk 2]
```diff
@@ -189,6 +204,8 @@ export class DataTableBodyComponent implements OnInit, OnDestroy {
   @Input() summaryRow: boolean;
   @Input() summaryPosition: string;
   @Input() summaryHeight: number;
+  @Input() rowDraggable: boolean;
```
Comment (kfenner) [inline projects/ngx-datatable/src/lib/components/body/body.component.ts:207]: Does it make sense to have this on a global level? Shouldn't this be on every individual row?

File: projects/ngx-datatable/src/lib/components/body/body.component.ts [hunk 3]
```diff
@@ -886,4 +905,64 @@ export class DataTableBodyComponent implements OnInit, OnDestroy {
   onTreeAction(row: any) {
     this.treeAction.emit({ row });
   }
+
+  allowDrop(event: DragEvent, dropRow) {
+    if (this.rowDraggable) {
```
Comment (kfenner) [inline projects/ngx-datatable/src/lib/components/body/body.component.ts:910]: Isn't allowing rows to be dragged very different from dropping something (any kind of Element from outside the table) on a row? Aren't we mixing something here?

The idea is to have drag&drop support to be able to drop some other object from a different source onto a row. What does this have to do with whether the row is draggable or not?

File: projects/ngx-datatable/src/lib/components/datatable.component.ts [hunk 4]
```diff
@@ -494,6 +500,11 @@ export class DatatableComponent implements OnInit, DoCheck, AfterViewInit, After
    */
   @Output() treeAction: EventEmitter<any> = new EventEmitter();
 
+  /**
+   * Triggered when row is dragged and dropped on another row
```
Comment (kfenner) [inline projects/ngx-datatable/src/lib/components/datatable.component.ts:504]:
```suggestion
   * A row or another element is dragged and/or dropped on a row.
```

**Review by kfenner — COMMENTED** [review]
@chintankavathia Left some more comments

File: projects/ngx-datatable/src/lib/components/body/body.component.ts (hunk 3 above)
Comment (chintankavathia) [reply projects/ngx-datatable/src/lib/components/body/body.component.ts:910]: I assumed it to be internal dragg and drop. If we are also targeting external objects to be dropped then yes we need to change this.

File: projects/ngx-datatable/src/lib/components/body/body.component.ts (hunk 2 above)
Comment (chintankavathia) [reply projects/ngx-datatable/src/lib/components/body/body.component.ts]: changed to be at row level

File: projects/ngx-datatable/src/lib/components/body/body.component.ts (hunk 3 above)
Comment (chintankavathia) [reply projects/ngx-datatable/src/lib/components/body/body.component.ts]: Now it can be used drop external objects.

File: projects/ngx-datatable/src/lib/components/body/body.component.ts [hunk 5]
```diff
@@ -330,6 +347,8 @@ export class DataTableBodyComponent implements OnInit, OnDestroy {
   _offset: number;
   _pageSize: number;
   _offsetEvent = -1;
+  private _draggedRow: any;
```
Comment (fh1ch) [inline projects/ngx-datatable/src/lib/components/body/body.component.ts:350]: Let's create a separate group for private member variables:

```suggestion

  private _draggedRow: any;
```

File: projects/ngx-datatable/src/lib/types/drag-events.type.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
+export type DragEventType = 'drag' | 'dragend' | 'dragenter' | 'dragleave' |
+'dragover' | 'dragstart' | 'drop';
```
Comment (fh1ch) [inline projects/ngx-datatable/src/lib/types/drag-events.type.ts:2]: Content of the same type should be indented if wrapped on a new line:

```suggestion
  'dragover' | 'dragstart' | 'drop';
```

**Review by fh1ch — CHANGES_REQUESTED** [review]
@chintankavathia really nice work here, thanks a ton 🙇 

As already mentioned, I really like the current API design and think we can continue this route. Therefore, just a few stylistic comments from my end. Back to you 🏓

**Review by kfenner — APPROVED** [review]
@chintankavathia  Looks great now, all my concerns are addressed 💯 🥇

**Review by fh1ch — APPROVED** [review]
@chintankavathia again, thanks a ton for the work here 🙇 

I just did another pass without any new findings.

LGTM 👍
